src/libsyntax/json/tests.rs

```diff
@@ -0,0 +1,142 @@
+    pub line_start: u32,
+    pub column_start: u32,
+    pub line_end: u32,
+    pub column_end: u32,
+}
+
+/// Test the span yields correct positions in JSON.
+fn test_positions(code: &str, span: (u32, u32), expected_output: SpanTestData) {
+
+    let expected_output = TestData { spans: vec![expected_output] };
+
+    with_default_globals(|| {
+        let sm = Lrc::new(SourceMap::new(FilePathMapping::empty()));
+        sm.new_source_file(Path::new("test.rs").to_owned().into(), code.to_owned());
+
+        let output = Arc::new(Mutex::new(Vec::new()));
+        let je = JsonEmitter::new(
+            Box::new(Shared { data: output.clone() }),
+            None,
+            sm,
+            true,
+            HumanReadableErrorType::Short(ColorConfig::Never),
+            false
+        );
+
+        let span = Span::with_root_ctxt(BytePos(span.0), BytePos(span.1));
+        let handler = Handler::with_emitter(true, None, Box::new(je));
+        handler.span_err(span, "foo");
+
+        let bytes = output.lock().unwrap();
+        let actual_output = str::from_utf8(&bytes).unwrap();
+        let actual_output : TestData = decode(actual_output).unwrap();
+
+        println!("expected output:\n------\n{:#?}------", expected_output);
+        println!("actual output:\n------\n{:#?}------", actual_output);
+
+        assert!(expected_output == actual_output)
+    })
+}
+
+#[test]
+fn test_empty() {
+    test_positions(
+        " ",
+        (0, 1),
+        SpanTestData {
+            byte_start: 0,
+            byte_end: 1,
+            line_start: 1,
+            column_start: 1,
+            line_end: 1,
+            column_end: 2,
+        }
+    )
+}
+
+#[test]
+fn test_bom() {
+    test_positions(
+        "\u{feff} ",
+        (0, 1),
+        SpanTestData {
+            byte_start: 2,
+            byte_end: 3,
+            line_start: 1,
+            column_start: 1,
+            line_end: 1,
+            column_end: 2,
+        }
+    )
+}
+
+#[test]
+fn test_newlines() {
+    test_positions(
+        "\nmod foo;\nmod bar;\n",
+        (5, 12),
+        SpanTestData {
+            byte_start: 5,
+            byte_end: 12,
+            line_start: 2,
+            column_start: 5,
+            line_end: 3,
+            column_end: 3,
+        }
+    )
+}
+
+#[test]
+fn test_newlines_crlf() {
+    test_positions(
+        "\r\nmod foo;\r\nmod bar;\r\n",
+        (5, 12),
+        SpanTestData {
+            byte_start: 6,
+            byte_end: 14,
+            line_start: 2,
+            column_start: 5,
+            line_end: 3,
+            column_end: 3,
+        }
+    )
+}
+
+#[test]
+fn test_newlines_crlf_with_bom() {
+    test_positions(
+        "\u{feff}\r\nmod foo;\r\nmod bar;\r\n",
+        (5, 12),
+        SpanTestData {
+            byte_start: 8,
+            byte_end: 16,
+            line_start: 2,
+            column_start: 5,
+            line_end: 3,
+            column_end: 3,
+        }
+    )
+}
```

Comment:
I am not sure, but my gut feeling is that unit-tests that require context setup like `with_default_globals` are a liability long term, as they depend too much on the current architecture. For this reason, I think it's best to write an integration ui test here, like `ui/json-short.stderr`, but with some actual error and a file which starts with a BOM and contains `\r\n`.

Reply:
`src/libsyntax/tests.rs` already have a pile of tests with `with_default_globals` so I just adapted it from there. Personally I'd love to avoid it - but it seemed like `Arc` and/or `Mutex` did not work without it.

What I wanted to really avoid was relying on CRLFs being maintained in the files, but I just noted that there's already precedent for this in the form of `lexer-crlf-line-endings-string-literal-doc-comment.rs` that includes CRLF and gets special handling in `.gitattributes` to mark it as non-text to avoid git from normalizing the linebreaks.

I'll see if I can remove the `with_default_globals` from these tests _and_ I'll add an integration ui test. I felt that with these libsyntax unit tests it was easier to test for strict offsets (ie. spans right before, right at and right after the linebreak).

Reply:
No idea why I got the impression the default globals had something to do with Arc/Mutex. The tests are panicking within syntax_pos that.

I'd still leave these in for now since the same `with_default_globals` is already used in `libsyntax/tests.rs` and I still feel they assert in more controlled manner. I've added a ui test for them as well though.

Just to confirm, the byte positions should be start inclusive, end exclusive? As in...

```
     v <- span 0..1, bytes 3..4
<BOM>foo bar baz
         ^^^ <- span 4..7, bytes 7..10
```

Reply:
>Just to confirm, the byte positions should be start inclusive, end exclusive?

Yes